### PR TITLE
helm3: initial components 

### PR DESCRIPTION
**Pull request by lpuskas**
| Q               | A
| --------------- | ---
| Bug fix?        | no|
| New feature?    |yes
| API breaks?     | no|
| Deprecations?   |yes
| License         | Apache 2.0

### What's in this PR?
Initial port of pipeline's helm functionality to helm3 (helm version driven by a config flag)

### Why?
Long term all helm functionality to be moved to helm3 only

### CODE REVIEW

File: cmd/pipeline/main.go [hunk 1]
```diff
@@ -746,14 +750,39 @@ func main() {
...
+							helmFacade,
+							kitxendpoint.Combine(endpointMiddleware...),
+						)
+
+						helmdriver.RegisterReleaserHTTPHandlers(endpoints,
+							clusterRouter.PathPrefix("/deployments").Subrouter(),
+							kitxhttp.ServerOptions(httpServerOptions),
+						)
+
+						cRouter.POST("/deployments", gin.WrapH(router))
```
Comment (sagikazarmark) [inline cmd/pipeline/main.go:766]: Try to use `cRouter.Any`

File: cmd/pipeline/main.go [hunk 2]
```diff
@@ -746,14 +750,39 @@ func main() {
...
+						cRouter.GET("/deployments", gin.WrapH(router))
+						cRouter.GET("/deployments/:name", gin.WrapH(router))
```
Comment (sagikazarmark) [inline cmd/pipeline/main.go:768]: ☝️

File: internal/helm/envresolver.go [hunk 3]
```diff
@@ -60,38 +66,71 @@ type EnvResolver interface {
...
 		platform: true,
 	}, nil
 }
+
+// helm3EnvResolver helm env resolver to be used for resolving helm 3 environments
+type helm3EnvResolver struct {
+	delegate EnvResolver
+}
+
+func NewHelm3EnvResolver(delegate EnvResolver) EnvResolver {
```
Comment (sagikazarmark) [inline internal/helm/envresolver.go:110]: Reusing helm 2 code doesn't make much sense here as we want to delete it at some point.

File: internal/helm/helmdriver/transport_http.go [hunk 4]
```diff
@@ -23,54 +23,194 @@ import (
...
 	))
+
+	router.Methods(http.MethodGet).Path("/charts").Handler(kithttp.NewServer(
+		endpoints.ListCharts,
+		decodeListChartsHTTPRequest,
+		kitxhttp.ErrorResponseEncoder(encodeListChartsHTTPResponse, errorEncoder),
+		options...,
+	))
+
+	// TODO fix the path after migrating to h3 (use chartS instead of chart) - backwards  compatibility!
```
Comment (sagikazarmark) [inline internal/helm/helmdriver/transport_http.go:81]: 👍

File: internal/helm/helmdriver/transport_http.go [hunk 5]
```diff
@@ -23,54 +23,194 @@ import (
...
+	router.Methods(http.MethodGet).Path("/chart/{reponame}/{name}").Handler(kithttp.NewServer(
+		endpoints.GetChart,
+		decodeChartDetailsHTTPRequest,
+		kitxhttp.ErrorResponseEncoder(encodeChartDetailsHTTPResponse, errorEncoder),
+		options...,
+	))
+}
+
+func RegisterReleaserHTTPHandlers(endpoints Endpoints, router *mux.Router, options ...kithttp.ServerOption) {
```
Comment (sagikazarmark) [inline internal/helm/helmdriver/transport_http.go:90]: IMHO this should eventually be moved under `/helm/releases` and the deployment API should be retired.

File: internal/helm/service.go [hunk 6]
```diff
@@ -45,11 +45,37 @@ type Repository struct {
...
 // Service manages Helm chart repositories.
 type Service interface {
+	// helm repository management operations
+	repository
+
+	// release management operations
+	releaser
+
+	// chart related operations
+	charter
```
Comment (sagikazarmark) [inline internal/helm/service.go:73]: Interesting approach....I'm not sure I like it though. 😛 

The interface is harder to read this way.

You can split the implementation of the service into categories (eg. `service_release.go`).

File: internal/helm/releaser.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	// List retrieves  releases in a given namespace, eventually applies the passed in filters
+	ListReleases(ctx context.Context, organizationID uint, clusterID uint, filters interface{}, options Options) ([]Release, error)
+	// Get retrieves the release details for the given  release
+	GetRelease(ctx context.Context, organizationID uint, clusterID uint, releaseName string, options Options) (Release, error)
+	// Upgrade upgrades the given release
+	UpgradeRelease(ctx context.Context, organizationID uint, clusterID uint, release Release, options Options) error
+	// ReleaseStatus
+	ReleaseStatus(ctx context.Context, organizationID uint, clusterID uint, releaseName string, options Options) (string, error)
+	// ReleaseResources retrieves resources belonging to the release
+	ReleaseResources(ctx context.Context, organizationID uint, clusterID uint, release Release, options Options) ([]ReleaseResource, error)
```
Comment (sagikazarmark) [inline internal/helm/releaser.go:73]:
```suggestion
	// CheckRelease
	CheckRelease(ctx context.Context, organizationID uint, clusterID uint, releaseName string, options Options) (string, error)
	// GetReleaseResources retrieves resources belonging to the release
	GetReleaseResources(ctx context.Context, organizationID uint, clusterID uint, release Release, options Options) ([]ReleaseResource, error)
```

File: cmd/pipeline/main.go [hunk 8]
```diff
@@ -1252,3 +1276,34 @@ func createInternalAPIRouter(
...
 }
+
+// setupHelmFacade utility function for assembling the helm facade
+func setupHelmFacade(config configuration, db *gorm.DB, commonSecretStore common2.SecretStore,
+	clusterManager *cluster.Manager, logger helm.Logger) helm.Service {
+	repoStore := helmadapter.NewHelmRepoStore(db, logger)
+	secretStore := helmadapter.NewSecretStore(commonSecretStore, logger)
+	orgService := helmadapter.NewOrgService(logger)
+	validator := helm.NewHelmRepoValidator()
+	releaser := helmadapter.NewReleaser(logger)
```
Comment (pepov) [inline cmd/pipeline/main.go:1287]: v2 implementation is missing?

File: cmd/pipeline/main.go [hunk 9]
```diff
@@ -1252,3 +1276,34 @@ func createInternalAPIRouter(
 	internalGroup.PUT("/:orgid/clusters/:id/nodepools", clusterAPI.UpdateNodePools)
 	return internalRouter
 }
+
+// setupHelmFacade utility function for assembling the helm facade
+func setupHelmFacade(config configuration, db *gorm.DB, commonSecretStore common2.SecretStore,
```
[the remark this answers is not in the document]
Comment (lpuskas) [reply cmd/pipeline/main.go:1255]: It's used in 2 places for the time being

File: internal/helm/envresolver.go (hunk 3 above)
Comment (lpuskas) [reply internal/helm/envresolver.go:111]: sure, this seems overcomplicated (i didnt want to duplicat logic, will fix it)

File: internal/helm/helmdriver/transport_http.go (hunk 5 above)
Comment (lpuskas) [reply internal/helm/helmdriver/transport_http.go:97]: thx, will do it later / i wouldn't break the api at this point

File: internal/helm/service.go (hunk 6 above)
Comment (lpuskas) [reply internal/helm/service.go:76]: thx, will do it in a later PR

File: internal/helm/releaser.go (hunk 7 above)
Comment (lpuskas) [reply internal/helm/releaser.go:73]: thx, done

File: cmd/pipeline/main.go (hunk 8 above)
Comment (lpuskas) [reply cmd/pipeline/main.go:1261]: let's take it offline

File: cmd/pipeline/main.go (hunk 2 above)
Comment (lpuskas) [reply cmd/pipeline/main.go:728]: thx will do it later on

**Pull request merged**